### PR TITLE
Add onBuild to esbuild.serve

### PR DESCRIPTION
**Pull request by asendia**
An attempt to resolve https://github.com/evanw/esbuild/issues/1072

I use `metafile` during build to replace `index.js` to `index-[hash].js` in my html template, the code goes like this:
```
require('esbuild').build(buildOptions).catch((err) => {
  console.error(err);
  process.exit(1);
}).then((res) => {
  buildHtmlSync(res.metafile); // template replace based on metafile
});
```
But `metafile` is missing from `esbuild.serve` (CMIIW)

This PR will add optional `onBuild` callback into `esbuild.serve` first argument:
```
esbuild.serve({
  servedir: './out',
  onBuild: (res) => {
    buildHtmlSync(res.metafile);
  },
}, buildOptions).then((result) => {
...
```

Note: I am new to this, please give me pointers if I miss something 🙇🏼

### CODE REVIEW

File: pkg/api/serve_other.go [hunk 1]
```diff
@@ -53,6 +54,9 @@ func (h *apiHandler) build() BuildResult {
 				result := h.rebuild()
 				h.rebuild = result.Rebuild
 				build.result = result
+				if h.onBuild != nil {
```
Comment (asendia) [inline pkg/api/serve_other.go:57]: Not sure about this

File: lib/common.ts [hunk 2]
```diff
@@ -548,6 +550,15 @@ export function createChannel(streamIn: StreamIn): StreamOut {
           break;
         }
 
+        case 'serve-build': {
+          let callbacks = serveCallbacks.get(request.serveID);
+          let buildResult: types.BuildResult = { warnings: request.args.warnings }
```
Comment (asendia) [inline lib/common.ts:555]: Not sure if this is ok since during `build`, `warnings` is retrieved via `replaceDetailsInMessages(response!.errors, details);`

**Comment by palfrey** [comment]
@evanw What needs to be done to move this one forward? Would love to see this functionality in!

**Comment by palfrey** [comment]
> @evanw What needs to be done to move this one forward? Would love to see this functionality in!

@evanw Following up on this one again. Happy to do whatever work needs to be done to help here!

**Comment by evanw** [comment]
This underlying issue has been fixed as of [version 0.17.0](https://github.com/evanw/esbuild/releases/v0.17.0). Closing.

**Pull request closed without merging**